common/libs/scheduling/index.ts

```diff
@@ -0,0 +1,209 @@
+import BN from 'bn.js';
+import abi from 'ethereumjs-abi';
+import { toWei, Units, gasPriceToBase, Address } from '../units';
+import { toBuffer } from 'ethereumjs-util';
+import RequestFactory from './contracts/RequestFactory';
+
+const TIME_BOUNTY_MIN = new BN('1');
+
+export const EAC_SCHEDULING_CONFIG = {
+  DAPP_ADDRESS: 'https://app.chronologic.network',
+  SCHEDULE_GAS_LIMIT_FALLBACK: new BN('21000'),
+  SCHEDULE_GAS_PRICE_FALLBACK: 20, // Gwei
+  FEE: new BN('0'),
+  FUTURE_EXECUTION_COST: new BN('180000'),
+  SCHEDULING_GAS_LIMIT: new BN('1500000'),
+  WINDOW_SIZE_DEFAULT_TIME: 10,
+  WINDOW_SIZE_DEFAULT_BLOCK: 90,
+  TIME_BOUNTY_MIN,
+  TIME_BOUNTY_DEFAULT: TIME_BOUNTY_MIN,
+  TIME_BOUNTY_MAX: toWei('900', Units.ether.length - 1), // 900 ETH
+  SCHEDULE_TIMESTAMP_FORMAT: 'YYYY-MM-DD HH:mm:ss',
+  DEFAULT_SCHEDULING_METHOD: 'time'
+};
+
+export const EAC_ADDRESSES = {
+  KOVAN: {
+    blockScheduler: '0x1afc19a7e642761ba2b55d2a45b32c7ef08269d1',
+    requestFactory: '0x496e2b6089bde77293a994469b08e9f266d87adb',
+    timestampScheduler: '0xc6370807f0164bdf10a66c08d0dab1028dbe80a3'
+  }
+};
+
+export const calcEACFutureExecutionCost = (
+  callGas: BN,
+  callGasPrice: BN,
+  timeBounty: BN | null
+) => {
+  const totalGas = callGas.add(EAC_SCHEDULING_CONFIG.FUTURE_EXECUTION_COST);
+
+  if (!timeBounty) {
+    timeBounty = EAC_SCHEDULING_CONFIG.TIME_BOUNTY_MIN;
+  }
+
+  return timeBounty.add(EAC_SCHEDULING_CONFIG.FEE).add(totalGas.mul(callGasPrice));
+};
+
+export const calcEACEndowment = (
+  callGas: BN | null,
+  callValue: BN | null,
+  callGasPrice: BN | null,
+  timeBounty: BN
+) => {
+  callValue = callValue || new BN(0);
+
+  return callValue.add(
+    calcEACFutureExecutionCost(
+      callGas || EAC_SCHEDULING_CONFIG.SCHEDULE_GAS_LIMIT_FALLBACK,
+      callGasPrice || gasPriceToBase(EAC_SCHEDULING_CONFIG.SCHEDULE_GAS_PRICE_FALLBACK),
+      timeBounty
+    )
+  );
+};
+
+export const calcEACTotalCost = (
+  callGas: BN,
+  gasPrice: BN,
+  callGasPrice: BN | null,
+  timeBounty: BN | null
+) => {
+  if (!callGasPrice) {
+    callGasPrice = gasPriceToBase(EAC_SCHEDULING_CONFIG.SCHEDULE_GAS_PRICE_FALLBACK);
+  }
+
+  const deployCost = gasPrice.mul(EAC_SCHEDULING_CONFIG.SCHEDULING_GAS_LIMIT);
+
+  const futureExecutionCost = calcEACFutureExecutionCost(callGas, callGasPrice, timeBounty);
+
+  return deployCost.add(futureExecutionCost);
+};
+
+export const getScheduleData = (
+  toAddress: string,
+  callData: string | Buffer = '',
+  callGas: BN | null,
+  callValue: BN | null,
+  windowSize: BN | null,
+  windowStart: any,
+  callGasPrice: BN | null,
+  timeBounty: BN | null,
+  requiredDeposit: BN | null
+) => {
+  if (!requiredDeposit || requiredDeposit.lt(new BN(0))) {
+    requiredDeposit = new BN(0);
+  }
+
+  if (typeof callData === 'string') {
+    callData = toBuffer(callData);
+  }
+
+  if (
```

Comment:
it would be nice if large conditionals like these were commented in groups for maintainability